pkg/cluster/spec/config.go

```diff
@@ -0,0 +1,97 @@
+package spec
+
+import (
+	"bytes"
+	"reflect"
+
+	"github.com/BurntSushi/toml"
+	"github.com/pingcap/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// TomlConfig of an instance
+type TomlConfig struct {
+	mp map[string]interface{}
+}
+
+var _ yaml.Marshaler = &TomlConfig{}
+var _ yaml.Unmarshaler = &TomlConfig{}
+
+// UnmarshalYAML implements yaml.Unmarshal interface.
+func (c *TomlConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var value interface{}
+	err := unmarshal(&value)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	switch v := value.(type) {
+	case string:
+		err = toml.Unmarshal([]byte(v), &c.mp)
+		if err != nil {
+			return errors.AddStack(err)
+		}
+	case map[interface{}]interface{}:
+		var value map[string]interface{}
+		err := unmarshal(&value)
+		if err != nil {
+			return errors.AddStack(err)
+		}
+
+		c.mp, err = merge(value)
+		if err != nil {
+			return errors.AddStack(err)
+		}
+	default:
+		return errors.Errorf("unknown type: %v", reflect.TypeOf(value))
+
+	}
+
+	return nil
```

Comment:
We should unflatten the unshelled `map[string]interface{}` because we can guarantee all keys is `xxx.yyy` form. e.g:
```yaml
server_configs:
  pd:
    xxx.yyy: 1
    zzz:
      z1: 2
      z2: 3
```